Translate _d_arrayappend{T,cTX} to templates

Vild opened the pull request:
dmd PR: https://github.com/dlang/dmd/pull/9982

This PR adds templated version of `_d_arrayappend{T,cTX}`.

## Notes
- There are wrapper around each hook to fake the safety level of the real implementations to not break code.
- I only translated the entrypoints for the hooks and not the called functions. Because this would have taken took long. Future working can be done to rewrite these functions.
- `_d_arrayappend{c,w}d` were not translated because they did not use TypeInfo in their interface. But they were changed to use the new hooks.

thewilsonator commented:
This uses `scope`, this should also be using `return ref` if I am understating the code correctly. This may alleviate the need for trusted.

JinShil commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+ * Run postblit on `t` if it is a struct and needs it.
+ * Or if `t` is a array, run it on the children if they have a postblit.
+ */
+void __doPostblit(T)(auto ref T t)
```
Can this be `private` in this module?  Also, why the `__` prefix?

Vild replied:
Yes, I can add private. The `__` is because it called this inside rt.lifetime: https://github.com/dlang/druntime/blob/master/src/rt/lifetime.d#L701

JinShil commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+/**
+ * Append y[] to array x[]
+ */
+T[] _d_arrayappendT_(T)(ref T[] x, scope T[] y) @trusted
```
It looks like this is only used in one place.  What about embedding it in the caller as a nested function to make it more obvious that this is an implementation detail of that function?

Vild replied:
Good suggestion, will do.

JinShil commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+ * Extend an array by n elements.
+ * Caller must initialize those elements.
+ */
+T[] _d_arrayappendcTX_(T)(ref T[] px, size_t n) @trusted
```
Same for this.  I'm recommending that this be embedded as a nested function in its only caller to keep the implementations together.

JinShil commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+// These are needed to fake the safety level to `@safe pure nothrow`.
+// Without this code will break as these hooks where previously exported
+// as having this safety level.
+T[] _d_arrayappendT(T)(ref T[] x, scope T[] y) @trusted
```
Do we need to return `T[]`?

JinShil replied:
Is the `@trusted` on this declaration redundant?  It seems that since the call to `_d_arrayappendT_` is wrapped `@trusted.` the  `@trusted` attribute at entry point is not necessary.

thewilsonator replied:
> Do we need to return T[]?

For chaining?

Vild replied:
It returns `T[]` because the original hook returned `void[]`. I also just found a problem with the current implementation in dmd:
```d
int[] arr;
((arr ~= [1,2,3]) ~= 4) ~= 5;
```
This crashes in the backend for dmd (https://run.dlang.io/is/UZAIuX), while working as expected in ldc (`arr == [1, 2, 3, 4, 5]`).
So I should probably change this to return a `ref T[]`, to match ldc's behavior.

I think I still need `@trusted` for the outer functions as I get this error if I remove `@trusted` on `_d_arrayappendcT`.
```
std/uni.d(6165): Error: @safe function std.uni.Stack!(InversionList!(GcPolicy)).Stack.push cannot call @system function object._d_arrayappendcT!(InversionList!(GcPolicy))._d_arrayappendcT
```

JinShil commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+        pragma(mangle, mangle) mixin(func);
+    }
+
+    alias __arrayStart = externDFuncMangle!("_D2rt8lifetime12__arrayStartFNaNbS4core6memory8BlkInfo_ZPv", "extern(D) void * externDFuncMangle(BlkInfo info) nothrow pure;");
```
Looks like we have some structural problems in druntime, if this is required.  I don't know the solution at this time.

JinShil replied:
Can a lot of this logic be encapsulated in a function in `rt.lifetime` so most of this can be replaced with a single function call, and avoid having to expose so many internals of `rt.lifetime` to  here?

Vild replied:
One thing I could do is that I could call `rt.lifetime._d_arrayappendcTX` directly instead of porting it. As I'm not sure I have time to start refactoring that function for this 2 week-block.

jpf91 replied:
Maybe that's best for now. If the compiler instantiates the template (wrapper), we can always port the implementation later on.

JinShil commented:
> Should I keep the changes to _d_arrayappend{c,w}d for this PR?
> If those changes gets removed and this PR and dmds PR gets merged. The old code will still be used when assigning wchar and dchars to char[]. But then yet again, those changed do not really belong in this PR.
Not sure should that have been a prerequisite to this PR?

jpf91 commented:
>libphobos2.a(process.o):std/process.d:_D3std7process12__ModuleInfoZ: error: undefined reference to '_D3std8internal6memory12__ModuleInfoZ'
> I have not figured out what causes this to happen. I guess it is something related to the templates using internal imports, but I'm not sure. I would appreciate some help.

The really strange thing here and the reason this fails to link is that `std.internal.memory` is in phobos. 
Can you run `process.o` through objdump and see where exactly the symbol is referenced?

> _D4core3sys7openbsdQm5cdefs12__ModuleInfoZ

Also strange The `core` modules are not compiled into the library, so there's not ModuleInfo, TypeInfo, ... emitted for these. There's nothing in `cdefs.d` which would need ModuleInfo though and I wonder why it even pulls in OpenBSD code?

jacob-carlborg commented:
> Also strange The core modules are not compiled into the library, so there's not ModuleInfo, TypeInfo, ... emitted for these. 

I’ve run into this before. The C bindings are not being compiled. So the few macros that have been translated to functions are not compiled and causes linker errors.

jacob-carlborg commented:
> There's nothing in cdefs.d which would need ModuleInfo though and I wonder why it even pulls in OpenBSD code?

Perhaps compiling with `-vgc-ast` will reveal something.

Vild commented:
> > Should I keep the changes to _d_arrayappend{c,w}d for this PR?
> > If those changes gets removed and this PR and dmds PR gets merged. The old code will still be used when assigning wchar and dchars to char[]. But then yet again, those changed do not really belong in this PR.
> 
> I'm not sure. Should that have been a prerequisite to this PR?

I think I will remove that commit, and then apply it later when I do a clean-up PR.

rainers commented on src/object.d:
```diff
@@ -4956,3 +4956,315 @@ string _d_assert_fail(string comp, A, B)(A a, B b) @nogc @safe nothrow pure
+
+T[] _d_arrayappendcTXTrace(T)(string file, int line, string funcname, ref T[] px, size_t n) @trusted pure
+{
+    import rt.profilegc : accumulate;
```
Note that the implementation of the tracing functions has changed recently, see https://github.com/dlang/druntime/pull/2607

I wouldn't mind if the trace-functions are left unmodified, it's a half-baked feature anyway and slows down execution in accumulate.

Vild replied:
The problem is that the new entrypoints for the hooks are templates, so I need to manually write a trace entrypoint. It also need to be in object.d so the compiler can reference it from the lowering code. https://github.com/dlang/dmd/pull/9982/files#diff-0c1aa1ff99402eb87e11cc3fa1b80da0

Vild commented:
I've updated the code according to the comments, but the unittest will still fail. This time because when I changed the compiler to rewrite `arr ~= elem` to `_d_arrayappendcT(arr, elem)` it will call the postblit twice on elem, once before the hook is called, and once during the hook. Here I either need to figure out how to never make it call the postblit before calling the function, or revert to the previous behavior where the assignment step is done outside the hook.

Vild commented:
Update:
All the druntime unittests passes.
In phobos everything related to `CodepointSet` fails, like this one: https://github.com/dlang/phobos/blob/master/std/uni.d#L6916

JinShil commented on posix.mak:
```diff
@@ -176,6 +176,9 @@ $(DOCDIR)/core_sys_darwin_netinet_%.html : src/core/sys/darwin/netinet/%.d $(DMD
 $(DOCDIR)/rt_%.html : src/rt/%.d $(DMD)
 	$(DMD) $(DDOCFLAGS) -Df$@ project.ddoc $(DOCFMT) $<
 
+$(DOCDIR)/rt_array_array.html : src/rt/array/array.d $(DMD)
```
I think you mean `src/rt/array/append.d`.

Vild replied:
Thanks :)

JinShil commented on src/object.d:
```diff
@@ -49,6 +49,14 @@ public import rt.array.capacity: reserve;
+/// See $(REF capacity, rt,array,capacity)
+public import rt.array.capacity: capacity;
+
+/// See $(REF _d_arrayappendT, rt,array,append) and $(REF _d_arrayappendTTrace, rt,array,append)
```
You actually don't need to document these, because they are not called directly by the user.  `capacity.d` has functions that are called directly by the user, so it is an exception.  I'd like to add a `version(Core_Doc)` block to filter out the runtime hooks so they don't appear in the documentation.

Vild replied:
Okey, thanks.

It think it is better to do the `version(Core_Doc)` in a separate PR.

JinShil commented on src/object.d:
```diff
@@ -41,6 +41,7 @@ version (D_ObjectiveC) public import core.attribute : selector;
 public import rt.array.comparison : __cmp;
 public import rt.array.equality : __ArrayEq, __equals;
 public import rt.array.casting: __ArrayCast;
+public import rt.array.append : _d_arrayappendT, _d_arrayappendTTrace,_d_arrayappendcTX, _d_arrayappendcTXTrace;
```
Nit: space after comma

JinShil commented on src/rt/array/append.d:
```diff
@@ -0,0 +1,290 @@
+// release. It also cannot be core.internal.externDFunc'd because that causes recursive template instantiation.
+
+/// See $(REF _d_arrayappendcTX, rt,lifetime,_d_arrayappendcTX)
+private extern (C) byte[] _d_arrayappendcTX(const TypeInfo ti, ref byte[] px, size_t n);
```
I think I commented about this before, but I can't find my comment now.  I think this can be imported instead of being declared as an `extern`

Vild replied:
I don't think it can because it will be instantiated inside the template and rt.lifetime not part of a dmd installation.

JinShil replied:
Ok, it looks like we'll eventually need to change the DMD installation to make all this templating of druntime work, but that is out of scope of the task at hand.

JinShil commented on src/rt/array/append.d:
```diff
@@ -0,0 +1,290 @@
+{
+    import core.internal.traits : hasElaborateCopyConstructor;
+
+    static if (is(T == struct)) {
```
Curly brace on next line

JinShil commented on src/rt/array/append.d:
```diff
@@ -0,0 +1,290 @@
+        // This template will be instanciated even in betterC due to the lowering code.
+        // So we need to disallow the calling of this function if that happens.
+        version (D_BetterC)
+            assert(0, "_d_arrayappendcTX does not support betterC!");
```
What happens if you don't specifically have this check.  Won't the -betterC build just emit a generic error about not supporting `TypeInfo`?

JinShil replied:
Also, if you do choose to keep this message, I don't think it should specifically mention `_d_arrayappendcTX` because that never appears in the user's code.  I suggest something worded more as "betterC does not support appending arrays".  

And I think that it can be a static assert at the beginning of the `_d_arrayappendcTX` template so the user gets a compile-time assertion failure instead of a runtime assertion failure.

Vild replied:
So without this check `test19561.d` will fail with `Error: TypeInfo cannot be used with -betterC`, even if that expression will be intercepted by the CTFE later.

and it cannot be a static assert because that breaks the lowering code.
```
import/rt/array/append.d(121): Error: static assert:  "_d_arrayappendcTX does not support betterC!"
import/rt/array/append.d(35):        instantiated from here: _d_arrayappendcTX!(immutable(char))
import/core/internal/arrayop.d(320):        instantiated from here: _d_arrayappendT!(immutable(char))
import/core/internal/arrayop.d(45):        instantiated from here: initScalarVecs!(int, "+=")
import/object.d(3696):        instantiated from here: arrayOp!(int[], int, "+=")
```

jpf91 commented on src/rt/array/append.d:
```diff
@@ -0,0 +1,292 @@
+            else
+                auto ti = typeid(T[]);
+
+            // _d_arrayappendcTX takes the `px` as a ubyte[], but its length
```
nitpick: as a byte[]

jpf91 commented:
So as it seems we can't avoid the `D_BetterC` workaround right now, after changing the version condition to `D_TypeInfo`, is there anything left to be done here or is this ready?

Code LGTM.

Vild commented:
> [...] is there anything left to be done here or is this ready?
I did the last few fixes and I'm happy with it for now.
In the future I would like to add `nothrow` deducting, but as it still goes though the old hooks it is pretty hard to add right now.

Vild commented:
This PR needs to implements the same fixes as in https://github.com/dlang/druntime/pull/2667, before it should be merged

RazvanN7 commented:
@Vild Can you rebase this to get rid of the conflicts and see if we can get this in?

Vild commented:
> @Vild Can you rebase this to get rid of the conflicts and see if we can get this in?

This PR requires https://github.com/dlang/druntime/pull/2673.
So as soon as I fix that one and it gets merge I will rebase this one.

thewilsonator commented:
> and it gets merge

Done.

Vild commented:
This PR is ready to be merged
